# HOMEWORK 1

#### Arish Balasubramaniam CIN : 304455611
#### Ishankumar Patel            CIN : 305096160
#### Cheryl Maria Jose          CIN : 305052753

#### a) Read the dataset file Cancer.csv and store it in pandas DataFrame

##### Step1: Importing the sklearn class (machine learning algorithm) that you would like to use for modeling:

In [1]:
# Numpy is a library for performing advanced mathematical operations on Arrays
# Pandas is a powerful library to read and process data tables
import numpy as np
import pandas as pd

##### Step2: Set up the Feature Matrix and Label Vector:

In [2]:
# loading a CSV file from local device and store it in a pandas DataFrame:
# "read_csv" is a pandas function to read csv files from web or local device:

df = pd.DataFrame()
df = pd.read_csv('https://raw.githubusercontent.com/Arishb/CS5661-Datascience/master/Cancer_small.csv')

In [3]:
# displaying the DataFrame:
# you can also use print(df)
df


Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                  5                        1                         1   
1                  5                        4                         4   
2                  3                        1                         1   
3                  6                        8                         8   
4                  4                        1                         1   
5                  8                       10                        10   
6                  1                        1                         1   
7                  2                        1                         2   
8                  2                        1                         1   
9                  4                        2                         1   
10                 5                        3                         3   
11                 8                        7                         5   
12                 7                        4                         6   
13                10                        7                         7   
14                 7                        3                         2   
15                10                        5                         5   
16                 8                        4                         5   
17                 5                        2                         3   
18                10                        7                         7   
19                10                       10                        10   
20                 5                        4                         4   
21                 2                        5                         3   
22                10                        4                         3   
23                 6                       10                        10   
24                 5                        6                         5   
25                10                       10                        10   
26                 3                        7                         7   
27                 1                        1                         1   
28                 4                        1                         1   
29                 7                        8                         7   
..               ...                      ...                       ...   
120                8                        6                         5   
121                5                        8                         7   
122                2                        1                         1   
123                5                       10                        10   
124                4                        1                         1   
125                5                        3                         3   
126                1                        1                         1   
127                1                        1                         1   
128                6                        1                         1   
129                5                        8                         8   
130                8                        7                         6   
131                2                        1                         1   
132                1                        5                         8   
133               10                        5                         6   
134                5                        8                         4   
135                1                        2                         3   
136               10                       10                        10   
137                7                        5                        10   
138                5                        1                         1   
139                1                        1                         1   
140                3                        1                         1   
141                4          

In [4]:
# this line print the size of heart_s.data (heart feature matrix) 
#(303 samples and 10 features)
print(df.shape)

(150, 10)


In [5]:
# checking the dataset by printing every 10 lines:
df[0::10]

Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                  5                        1                         1   
10                 5                        3                         3   
20                 5                        4                         4   
30                 9                        5                         8   
40                 5                        3                         5   
50                 5                        1                         3   
60                 2                        2                         2   
70                 1                        1                         1   
80                10                        3                         5   
90                 1                        3                         1   
100                5                        3                         5   
110                5                        5                         5   
120                8                        6                         5   
130                8                        7                         6   
140                3                        1                         1   

     Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                    1                            2            1   
10                   3                            2            3   
20                   9                            2           10   
30                   1                            2            3   
40                   5                            3            3   
50                   1                            2            1   
60                   1                            1            1   
70                   1                            2            1   
80                   1                           10            5   
90                   2                            2            2   
100                  1                            8           10   
110                  6                            3           10   
120                  4                            3           10   
130                  4                            4           10   
140                  1                            2            1   

     Bland_Chromatin  Normal_Nucleoli  Mitoses  Malignant_Cancer  
0                  3                1        1                 0  
10                 4                4        1                 1  
20                 5                6        1                 1  
30                 2                1        5                 1  
40                 4               10        1                 1  
50                 2                1        1                 0  
60                 7                1        1                 0  
70                 3                1        1                 0  
80                 3               10        2                 1  
90                 5                3        2                 0  
100                5                3        1                 1  
110                3                1        1                 1  
120                6                1        1                 1  
130                5                1        1                 1  
140                3                1        1                 0

##### Step3: Create Feature Matrix and Label Vector

In [6]:
# Creating the Feature Matrix

# create a python list of feature names that would like to pick from the dataset:
feature_cols = ['Clump_Thickness','Uniformity_of_Cell_Size','Uniformity_of_Cell_Shape','Marginal_Adhesion','Single_Epithelial_Cell_Size','Bare_Nuclei','Bland_Chromatin','Normal_Nucleoli','Mitoses']

# use the above list to select the features from the original DataFrame
X = df[feature_cols]  

In [7]:
# print the first 5 rows
X.head()

Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   
3                6                        8                         8   
4                4                        1                         1   

   Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                  1                            2            1   
1                  5                            7           10   
2                  1                            2            2   
3                  1                            3            4   
4                  3                            2            1   

   Bland_Chromatin  Normal_Nucleoli  Mitoses  
0                3                1        1  
1                3                2        1  
2                3                1        1  
3                3                7        1  
4                3                1        1

In [8]:
# checking the size of Feature Matix X:
print(X.shape)

(150, 9)


In [9]:
# select a Series of labels (the last column) from the DataFrame
y = df['Malignant_Cancer']

In [10]:
# checking the label vector by printing every 10 values
y[::10]

0      0
10     1
20     1
30     1
40     1
50     0
60     0
70     0
80     1
90     0
100    1
110    1
120    1
130    1
140    0
Name: Malignant_Cancer, dtype: int64

#### b) Split the dataset into testing and training set uisng Sklearn functions and the following parameter : test_size=0.3 and random_state=2

##### Step4 : Split the dataset into training and testing set

In [11]:
# Randomly splitting the original dataset into training set and testing set
# The function"train_test_split" from "sklearn.cross_validation" library performs random splitting.
# "test_size=0.3" means that pick 30% of data samples for testing set, and the rest (70%) for training set.

from sklearn.cross_validation import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=2)

In [12]:
# print the size of the traning set:
print(X_train.shape)
print(y_train.shape)

(105, 9)
(105,)


In [13]:
# print the size of the testing set:
print(X_test.shape)
print(y_test.shape)

(45, 9)
(45,)


In [14]:
print(X_test)
print('\n')
print(y_test)

     Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
6                  1                        1                         1   
3                  6                        8                         8   
113                5                        6                         7   
12                 7                        4                         6   
24                 5                        6                         5   
129                5                        8                         8   
25                10                       10                        10   
108                7                        2                         4   
128                6                        1                         1   
45                10                        4                         2   
48                 8                        3                         8   
42                 9                       10                        10   
35                10     

#### c) Use 'Decision Tree' to predict the cancer based on the training and testing sets that is in part (b). Calculate and report the accuracy of our classifier

##### Step5 : Import Decision tree classifier

In [15]:
# The following line will import DecisionTreeClassifier Classes
# DecisionTreeClassifier" is the name of a "sklearn class" to perform Decision Tree based Classification 

from sklearn.tree import DecisionTreeClassifier

##### Step6 : Instantiate an "object" of DecisionTreeClassifier "class".

In [16]:
# "my_DecisionTree" is instantiated as an "object" of DecisionTreeClassifier "class".

my_DecisionTree = DecisionTreeClassifier(random_state=2)

##### Step7: Training ONLY on the training set

In [17]:
# We can use the method "fit" of the object "my_DecisionTree" along with training dataset and labels to train the model.

my_DecisionTree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=2, splitter='best')

##### Step8: Testing on the testing set

In [18]:
y_predict_dt = my_DecisionTree.predict(X_test)

In [19]:
print(y_predict_dt)

[0 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 0 1 0 0 0 0 0 1 0 0 0 1 0 1 1 0 1 0 0
 1 1 0 1 1 0 1 1]


##### Step9: Accuracy Evaluation:
##### After prediction, we can now compare the "predicted labels" for the Testing Set with its "actual labels" to evaluate the accuracy of our KNN Classifier!

In [20]:
# Function "accuracy_score" from "sklearn.metrics" will perform the element-to-element comparision and returns the 
# portion of correct predictions:

from sklearn.metrics import accuracy_score

score_dt = accuracy_score(y_test, y_predict_dt)
print(score_dt)

0.866666666667


###### Accuracy for Decision Tree Classifier is : 0.866666666667

#### e) Use scikit-learn "Adaboost" classifier to predict Cancer based on the training and testing datasets. Calculate and report the accuracy of the classifier

##### Step10 : Import AdaBoost classifier

In [21]:
# The following line will import AdaBoost Classes
# AdaBoost Classifier" is the name of a "sklearn class" to perform Decision Tree based Classification

from sklearn.ensemble import AdaBoostClassifier

##### Step11 : Instantiate an "object" of AdaBoostClassifier "class"

In [22]:
my_AdaBoost = AdaBoostClassifier(n_estimators=19,random_state=2 )

##### Step12 : Repeat Step 7,8 and 9 for AdaBoostClassifier

In [23]:
my_AdaBoost.fit(X_train,y_train)

AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=1.0, n_estimators=19, random_state=2)

In [24]:
y_predict_AdB = my_AdaBoost.predict(X_test)

In [25]:
print(y_predict_AdB)

[0 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 0 0 1 0 0 0 0 1 1 0 0 0 1 0 1 1 0 1 0 1
 1 1 0 1 1 0 1 1]


##### Accuracy Evaluation for AdaBoostClassifier :

In [26]:
score_AdB = accuracy_score(y_test, y_predict_AdB)
print(score_AdB)

0.933333333333


###### Accuracy for AdaBoost Classifier is : 0.933333333333


#### f) Use scikit-learn "RandomForest" classifier to predict Cancer based on the training and testing datasets. Calculate and report the accuracy of the classifier

##### Step13 : Import RandomForest classifier

In [27]:
# The following line will import RandomForest Classes
# RandomForest Classifier" is the name of a "sklearn class" to perform Decision Tree based Classification

from sklearn.ensemble import RandomForestClassifier

#### Step14 : Instantiate an "object" of RandomForestClassifier "class"

In [28]:
my_RandomForest = RandomForestClassifier(n_estimators=19,bootstrap=True,random_state=2 )

##### Step15 : Repeat Step 7,8 and 9 for RandomForestClassifier

In [29]:
my_RandomForest.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=19, n_jobs=1,
            oob_score=False, random_state=2, verbose=0, warm_start=False)

In [30]:
y_predict_RF = my_RandomForest.predict(X_test)

In [31]:
print(y_predict_RF)

[0 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 0 1 0 1 0 0 0 0 1 1 0 0 0 1 0 1 1 0 1 1 1
 1 1 0 1 1 0 1 1]


##### Accuracy Evaluation for RandomForest Classifier :

In [264]:
score_RF = accuracy_score(y_test, y_predict_RF)
print(score_RF)

0.955555555556


###### Accuracy for Random Forest Classifier is : 0.955555555556

#### d) Write your own bagging method

In [32]:
from sklearn.utils import resample


In [33]:
###Define a new base decision tree classifier
 
Base_DecisionTree = DecisionTreeClassifier(random_state=2)

In [34]:
results = []

for k in range(0, 19):
    
    #1) Make a bootstrap sample of the original dataset
    bootstrap_size = 0.8*len(X_train)
    #2) Command to generate a random bootstrap dataset
    X_train_sample = resample(X_train, n_samples=bootstrap_size, random_state=k,replace = True)
    y_train_sample = resample(y_train, n_samples=bootstrap_size, random_state=k,replace = True)
    #3) Train the new base decision tree classifier
    Base_DecisionTree.fit(X_train_sample,y_train_sample)
    #4) Test this "base classifier" on the original testing dataset and save the predictions 
    y_predict_bdt = Base_DecisionTree.predict(X_test)
    print(y_predict_bdt)   
    results.append(y_predict_bdt)
    
#5) Save the result into a matrix    
print(results) 
mat = np.matrix(results)
#6) Print the matrix and size of a matrix
print(mat)
mat.shape
#7) Transpose the matrix so that we get a (45,19) matrix 
mat2 = mat.transpose()
#8) Print the matrix and size of a the new matrix
print(mat2)
mat2.shape


[0 1 1 1 1 1 1 0 0 1 1 1 1 1 0 0 0 1 0 1 0 0 0 1 0 1 0 0 0 1 0 1 1 0 1 1 0
 0 1 0 1 1 0 1 1]
[0 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 0 1 0 0 0 0 1 1 0 0 0 1 0 1 1 0 1 0 1
 1 1 0 1 1 0 1 1]
[0 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 0 1 0 0 0 0 0 1 0 0 0 1 0 1 1 0 1 0 0
 1 1 0 1 1 0 1 1]
[0 1 1 1 1 1 1 1 0 0 1 1 1 1 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 0 0 1 1 0 1 0 1
 0 1 0 1 1 0 1 1]
[0 1 1 0 0 1 0 1 0 0 1 1 1 1 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 0 0 1 1 0 0 0 1
 0 1 0 1 1 0 1 1]
[0 1 1 0 0 1 0 1 0 1 1 0 1 1 1 1 0 1 0 1 0 0 0 1 1 1 0 0 0 1 0 1 1 0 1 1 1
 1 1 0 1 1 0 1 1]
[0 1 1 1 1 1 1 1 0 1 1 1 1 1 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 1 0 1 1 0 1 1 1
 1 1 0 1 1 0 1 1]
[0 1 1 1 0 1 1 1 0 0 1 1 1 1 1 1 0 1 0 1 0 0 0 0 0 1 0 0 0 1 0 1 1 0 1 0 0
 1 1 0 1 1 0 1 1]
[0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 0 1 0 1 0 0 0 0 1 1 0 0 0 1 0 1 1 0 1 0 1
 1 1 0 1 1 0 1 1]
[0 1 1 0 0 1 0 0 0 1 1 1 1 1 1 1 0 1 0 1 0 0 0 1 1 1 0 0 0 0 0 1 1 0 0 0 1
 1 0 0 1 1 0 1 1]
[0 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1 0 1 1 0 1

/Users/cherylmariajose/anaconda/lib/python3.5/site-packages/sklearn/utils/__init__.py:258: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  indices = random_state.randint(0, n_samples, size=(max_n_samples,))


(45, 19)

In [35]:
#9)Take the sum of each row and store it in an array
sum_of_col = np.sum(mat2, axis=1)
print (sum_of_col)
#10)Store the length of the array in a variable
length=len(sum_of_col)

[[ 1]
 [19]
 [19]
 [15]
 [14]
 [19]
 [16]
 [16]
 [ 0]
 [ 7]
 [18]
 [18]
 [19]
 [19]
 [12]
 [10]
 [ 0]
 [11]
 [ 0]
 [17]
 [ 0]
 [ 0]
 [ 0]
 [ 8]
 [11]
 [19]
 [ 0]
 [ 0]
 [ 0]
 [15]
 [ 0]
 [19]
 [19]
 [ 0]
 [17]
 [ 8]
 [11]
 [12]
 [17]
 [ 0]
 [19]
 [19]
 [ 0]
 [19]
 [19]]


In [36]:
#11)Using the length of the array initialize a for loop 
res=[]
#12)Initialize a for loop to check if the sum of each column is lesser than 9
#if it is lesser than 9 then assign the value of 0 to a variable
#append these values into another list
for i in range(0,length):
    if sum_of_col[i] < 9:
                         m = 0
    else:
        m=1
    res.append(m)  
print(res)

[0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1]


In [37]:
#14) Check the accuracy for our bagging procedure
score_bag = accuracy_score(y_test, res)
print(score_bag)

0.911111111111


###### Accuracy for our bagging method is : 0.9111

In [271]:
#predict = stats.mode(mat2,axis=0)
#print(predict)